test/test/disconnected.cpp

```diff
@@ -122,3 +123,73 @@ TEST_CASE("disconnected")
         WaitForSingleObject(signal.get(), INFINITE);
     }
 }
+
+// Custom action to simulate an out-of-process server that crashes before it can complete.
+struct non_agile_abandoned_action : implements<non_agile_abandoned_action, IAsyncAction, IAsyncInfo, non_agile>
+{
+    non_agile_abandoned_action(void* event_handle) : m_awaited(event_handle) {}
+
+    void Completed(AsyncActionCompletedHandler const& handler) {
+        m_handler = handler;
+        // Tell the test to disconnect the IAsyncAction, which simulates the server crash.
+        SetEvent(m_awaited);
+    }
+    auto Completed() { return m_handler; }
+    void GetResults() {}
+
+    auto Id() { return 0U; }
+    auto Status() { return AsyncStatus::Completed; }
+    auto ErrorCode() { return hresult(0); }
+    void Cancel() {}
+    void Close() {}
+
+    AsyncActionCompletedHandler m_handler;
+    HANDLE m_awaited;
+};
+
+namespace
+{
+    template<typename TLambda>
+    void InvokeInContext(IContextCallback* context, TLambda&& lambda)
+    {
+        ComCallData data;
+        data.pUserDefined = &lambda;
+        check_hresult(context->ContextCallback([](ComCallData* data) -> HRESULT
+            {
+                auto& lambda = *reinterpret_cast<TLambda*>(data->pUserDefined);
+                lambda();
+                return S_OK;
+            }, &data, IID_ICallbackWithNoReentrancyToApplicationSTA, 5, nullptr));
+    }
+
+    fire_and_forget disconnect_on_signal(com_ptr<IContextCallback> context, void* signal)
+    {
+        co_await resume_on_signal(signal);
+        InvokeInContext(context.get(), []()
+            {
+                // This disconnects the IAsyncAction, simulating a server crash.
+                CoDisconnectContext(INFINITE);
+            });
+    }
+}
+
+TEST_CASE("disconnected,action")
+{
+    auto private_context = create_instance<IContextCallback>(CLSID_ContextSwitcher);
+    handle signal{ CreateEventW(nullptr, true, false, nullptr) };
+    disconnect_on_signal(private_context, signal.get());
+
+    agile_ref<IAsyncAction> action;
+    InvokeInContext(private_context.get(), [&]()
+        {
+            action = make<non_agile_abandoned_action>(signal.get());
+        });
+
+    auto result = [](IAsyncAction action) -> IAsyncAction
+        {
+            co_await action;
+            co_return;
```

Comment:
The `co_return` is superfluous.

Reply:
I have a vague memory that an earlier version of the coroutine spec required `co_return` for all coroutines, so I get kind of paranoid about it. Looks like the current spec allows `co_return` to be elided for `return_void` promises.